cpp/src/join/mixed_join.cu

```diff
@@ -0,0 +1,578 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/ast/detail/expression_parser.hpp>
+#include <cudf/ast/expressions.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/join.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/table/table_view.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/span.hpp>
+#include <join/hash_join.cuh>
+#include <join/join_common_utils.cuh>
+#include <join/join_common_utils.hpp>
+#include <join/mixed_join_kernels.cuh>
+
+#include <rmm/cuda_stream_view.hpp>
+
+#include <optional>
+#include <utility>
+
+#include <cstdio>
+
+namespace cudf {
+namespace detail {
+
+// TODO: Clearly document that it is the user's responsibility to match the
+// null operators in the predicate and the compare_nulls parameter.
+// TODO: Clearly document that the parameters for the size computation must
+// match exactly the corresponding call to the join operation. This
+// documentation should be added to the nested loop conditional join as well.
+// TODO: Clearly document failure cases.
+std::pair<std::unique_ptr<rmm::device_uvector<size_type>>,
+          std::unique_ptr<rmm::device_uvector<size_type>>>
+mixed_join(table_view const& left_equality,
+           table_view const& right_equality,
+           table_view const& left_conditional,
+           table_view const& right_conditional,
+           ast::expression const& binary_predicate,
+           // TODO: Figure out how to propagate compare_nulls to the query of
+           // the hash table in device code.
+           null_equality compare_nulls,
+           join_kind join_type,
```

Comment:
Since `join_type` is a constant expression, how about passing it via template?

Reply:
I decided against that mostly to reduce compile-time. Mixing `cuco::static_multimap` with cudf's AST machinery makes this TU _extremely_ slow to compile. Any change to `mixed_join.cu` or `mixed_join_kernels.cuh` ends up taking close to 10 mins to recompile. I didn't want to make that worse by templating on the `join_kind` since from my benchmarks of the conditional join kernel it doesn't really help speed up the kernels (those checks are cheap).

Reply:
Unless other people have a strong opinion on this, I think it's not worth trying then.